File: Tutorials/Codes/Introduction_to_Robotics_Course/Chapter7/Lesson5/Chapter7_Lesson5.m
fs = 100; Ts = 1/fs; N = 200;
t = (0:N-1)*Ts;

x = sin(2*pi*3*t);
sigma_n = 0.05;
sigma_b = 1e-4;

% Noise
n = sigma_n*randn(size(t));

% Drift as random walk
b = zeros(size(t));
for k=2:N
    b(k) = b(k-1) + sigma_b*randn();
end

y = x + n + b;

% Quantization
B = 10; ymin=-2; ymax=2;
Delta = (ymax-ymin)/2^B;
yq = Delta*round(y/Delta);

% Compare measured quantization variance to Delta^2/12
eq = yq - y;
fprintf('Delta=%g, Var(eq)=%g, Theoretical=%g\n', Delta, var(eq), Delta^2/12);

% Bias estimate via mean on first 20 samples
b_hat = mean(yq(1:20));
y_comp = yq - b_hat;

File: Tutorials/Codes/Introduction_to_Robotics_Course/Chapter7/Lesson5/Chapter7_Lesson5.java
import java.util.Random;

public class SensorModel {
    public static void main(String[] args){
        double fs = 100.0, Ts = 1.0/fs;
        int N = 200;
        double[] x = new double[N];
        double[] b = new double[N];
        double[] yq = new double[N];

        Random rng = new Random(0);

        double sigmaN = 0.05;
        double sigmaB = 1e-4;

        int Bbits = 10;
        double ymin = -2.0, ymax = 2.0;
        double Delta = (ymax - ymin) / (1 << Bbits);

        for(int k=0; k<N; k++){
            double t = k*Ts;
            x[k] = Math.sin(2*Math.PI*3*t);
            if(k==0) b[k]=0; 
            else b[k] = b[k-1] + sigmaB*rng.nextGaussian();

            double y = x[k] + b[k] + sigmaN*rng.nextGaussian();
            yq[k] = Delta * Math.round(y/Delta);
        }

        // Bias estimate from initial window
        double bHat = 0;
        for(int k=0; k<20; k++) bHat += yq[k];
        bHat /= 20.0;

        System.out.println("Delta=" + Delta + ", bHat=" + bHat);
    }
}
